gateway-service-impl/src/main/java/org/hypertrace/gateway/service/EntityTypesProvider.java

```diff
@@ -0,0 +1,64 @@
+package org.hypertrace.gateway.service;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import lombok.NonNull;
+import org.hypertrace.core.grpcutils.context.ContextualKey;
+import org.hypertrace.gateway.service.common.util.EntityTypeServiceClient;
+import org.hypertrace.gateway.service.common.util.EntityTypeServiceV2Client;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class EntityTypesProvider {
+  private static final Logger LOG = LoggerFactory.getLogger(EntityTypesProvider.class);
+  private static final int DEFAULT_CACHE_SIZE = 1024;
+  private static final int DEFAULT_EXPIRE_DURATION_MIN = 30; // 30 minutes
+
+  // TenantId to Entity types cache
+  private final LoadingCache<ContextualKey<Void>, Set<String>> tenantIdToEntityTypesCache;
+
+  public EntityTypesProvider(
+      EntityTypeServiceClient entityTypeServiceClient,
+      EntityTypeServiceV2Client entityTypeServiceV2Client) {
+    CacheLoader<ContextualKey<Void>, Set<String>> cacheLoader =
+        new CacheLoader<>() {
+          @Override
+          public Set<String> load(@NonNull ContextualKey<Void> contextualKey) {
+            Set<String> entityTypesSet = new TreeSet<>(String.CASE_INSENSITIVE_ORDER);
+            entityTypeServiceClient
+                .getAllEntityTypes(contextualKey.getContext())
+                .forEach(entityType -> entityTypesSet.add(entityType.getName()));
+            entityTypeServiceV2Client
+                .getAllEntityTypes(contextualKey.getContext())
+                .forEach(entityType -> entityTypesSet.add(entityType.getName()));
+            return entityTypesSet;
+          }
+        };
+
+    tenantIdToEntityTypesCache =
+        CacheBuilder.newBuilder()
+            .maximumSize(DEFAULT_CACHE_SIZE)
+            .expireAfterWrite(DEFAULT_EXPIRE_DURATION_MIN, TimeUnit.MINUTES)
+            .build(cacheLoader);
+  }
+
+  public Set<String> getEntityTypes(ContextualKey<Void> contextualKey) {
```

Comment:
Accept a context rather than a contextual key. By asking for a key, you're forcing the caller to decide how the cache should be determined which is an implementation detail to this provider (as is the fact that it caches at all).